coderd/workspaceagents_test.go

```diff
@@ -509,6 +512,118 @@ func TestWorkspaceAgentClientCoordinate_BadVersion(t *testing.T) {
 	require.Equal(t, "version", sdkErr.Validations[0].Field)
 }
 
+func TestWorkspaceAgentClientCoordinate_ResumeToken(t *testing.T) {
+	t.Parallel()
+
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+
+	// We block DERP in this test to ensure that even if there's no direct
+	// connection, no shenanigans happen with the peer IDs on either side.
+	dv := coderdtest.DeploymentValues(t)
+	err := dv.DERP.Config.BlockDirect.Set("true")
+	require.NoError(t, err)
+	client, closer, api := coderdtest.NewWithAPI(t, &coderdtest.Options{
+		DeploymentValues: dv,
+	})
+	defer closer.Close()
+	user := coderdtest.CreateFirstUser(t, client)
+
+	// Change the DERP mapper to our custom one.
+	var currentDerpMap atomic.Pointer[tailcfg.DERPMap]
+	originalDerpMap, _ := tailnettest.RunDERPAndSTUN(t)
+	currentDerpMap.Store(originalDerpMap)
+	derpMapFn := func(_ *tailcfg.DERPMap) *tailcfg.DERPMap {
+		return currentDerpMap.Load().Clone()
+	}
+	api.DERPMapper.Store(&derpMapFn)
+
+	// Start workspace a workspace agent.
+	r := dbfake.WorkspaceBuild(t, api.Database, database.Workspace{
+		OrganizationID: user.OrganizationID,
+		OwnerID:        user.UserID,
+	}).WithAgent().Do()
+
+	agentCloser := agenttest.New(t, client.URL, r.AgentToken)
+	resources := coderdtest.AwaitWorkspaceAgents(t, client, r.Workspace.ID)
+	agentID := resources[0].Agents[0].ID
+
+	// Create a new "proxy" server that we can use to kill the connection
+	// whenever we want.
+	l, err := netListenDroppable("tcp", "localhost:0")
+	require.NoError(t, err)
+	defer l.Close()
+	srv := &httptest.Server{
+		Listener: l,
+		//nolint:gosec
+		Config: &http.Server{Handler: api.RootHandler},
+	}
+	srv.Start()
+	proxyURL, err := url.Parse(srv.URL)
+	require.NoError(t, err)
+	proxyClient := codersdk.New(proxyURL)
+	proxyClient.SetSessionToken(client.SessionToken())
+
+	// Connect from a client.
+	conn, err := func() (*workspacesdk.AgentConn, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel() // Connection should remain open even if the dial context is canceled.
+
+		return workspacesdk.New(proxyClient).
```

Comment:
This test _should_ be a unit test of the `workspaceAgentClientCoordinate` API handler, but instead it's complex integration test that includes the `workspacesdk`, `tailnet`, and `agent`.  I think we should wait on the integration test here until we can assert what we actually care about: that the client and agent don't lose connectivity if they're directly connected and `coderd` goes down.

A small set of unit tests should mock / fake the `api.TailnetClientService` and then

 1. send a request with a valid resume_token, and assert that we call the tailnet service with the encoded Peer ID.
 2. send a request with an invalid resume_token, and assert that we fail the request in the appropriate way
 3. send a request with no resume_token and assert that we call the tailnet service with a non-nil Peer ID.
 
 As is, this test has a lot of moving pieces to, essentially, assert that the PeerID gets reused.